pkg/controller/oam/applicationconfiguration/applicationconfiguration.go

```diff
@@ -0,0 +1,315 @@
+		r.record.Event(ac, event.Warning(reasonCannotRenderComponents, err))
+		ac.Status.SetConditions(v1alpha1.ReconcileError(errors.Wrap(err, errRenderComponents)))
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, ac), errUpdateAppConfigStatus)
+	}
+	log.Debug("Successfully rendered components", "workloads", len(workloads))
+	r.record.Event(ac, event.Normal(reasonRenderComponents, "Successfully rendered components", "workloads", strconv.Itoa(len(workloads))))
+
+	if err := r.workloads.Apply(ctx, workloads); err != nil {
+		log.Debug("Cannot apply components", "error", err, "requeue-after", time.Now().Add(shortWait))
+		r.record.Event(ac, event.Warning(reasonCannotApplyComponents, err))
+		ac.Status.SetConditions(v1alpha1.ReconcileError(errors.Wrap(err, errApplyComponents)))
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, ac), errUpdateAppConfigStatus)
+	}
+	log.Debug("Successfully applied components", "workloads", len(workloads))
+	r.record.Event(ac, event.Normal(reasonApplyComponents, "Successfully applied components", "workloads", strconv.Itoa(len(workloads))))
+
+	// Kubernetes garbage collection will (by default) reap workloads and traits
+	// when the appconfig that controls them (in the controller reference sense)
+	// is deleted. Here we cover the case in which a component or one of its
+	// traits is removed from an extant appconfig.
+	for _, e := range r.gc.Eligible(ac.GetNamespace(), ac.Status.Workloads, workloads) {
+		// https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable
+		e := e
+
+		log := log.WithValues("kind", e.GetKind(), "name", e.GetName())
+		record := r.record.WithAnnotations("kind", e.GetKind(), "name", e.GetName())
+
+		if err := r.client.Delete(ctx, &e); resource.IgnoreNotFound(err) != nil {
+			log.Debug("Cannot garbage collect component", "error", err, "requeue-after", time.Now().Add(shortWait))
+			record.Event(ac, event.Warning(reasonCannotGGComponents, err))
+			ac.Status.SetConditions(v1alpha1.ReconcileError(errors.Wrap(err, errGCComponent)))
+			return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, ac), errUpdateAppConfigStatus)
+		}
+		log.Debug("Garbage collected resource")
+		record.Event(ac, event.Normal(reasonGGComponent, "Successfully garbage collected component"))
+	}
+
+	ac.Status.Workloads = make([]v1alpha2.WorkloadStatus, len(workloads))
+	for i := range workloads {
+		ac.Status.Workloads[i] = workloads[i].Status()
+	}
+
+	ac.Status.SetConditions(v1alpha1.ReconcileSuccess())
+	return reconcile.Result{RequeueAfter: longWait}, errors.Wrap(r.client.Status().Update(ctx, ac), errUpdateAppConfigStatus)
+}
+
+// A Workload produced by an OAM ApplicationConfiguration.
+type Workload struct {
+	// ComponentName that produced this workload.
+	ComponentName string
+
+	// A Workload object.
+	Workload *unstructured.Unstructured
+
```

Comment:
I am not sure what is a "workload". We have `workloadDefinition` and a `component` is an instance of that. Also, why is the name of the workload here is called `ComponentName`?

Reply:
I touched on that a little in https://github.com/crossplane/crossplane/pull/1321#discussion_r387402721. The `Workload` has a `ComponentName` to indicate which component produced/rendered the workload.